packages/react-components/react-dialog/Spec.md

```diff
@@ -1,63 +1,524 @@
+
+```tsx
+type DialogSlots = {
+  /**
+   * The dialog element itself
+   */
+  root: Slot<'div'>;
+  /**
+   * Dimmed background of dialog.
+   * The default overlay is rendered as a `<div>` with styling.
+   * This slot expects a `<div>` element which will replace the default overlay.
+   * The overlay should have `aria-hidden="true"`.
+   */
+  overlay?: Slot<'div'>;
+  /**
+   * The root element of the Dialog right after Portal.
+   */
+  root: Slot<'div'>;
+};
+
+type DialogProps = ComponentProps<DialogSlots> & {
+  /**
+   * Dialog variations.
+   *
+   * `modal`: When this type of dialog is open, the rest of the page is dimmed out and cannot be interacted with. The tab sequence is kept within the dialog and moving the focus outside the dialog will imply closing it. This is the default type of the component.
+   *
+   * `non-modal`: When a non-modal dialog is open, the rest of the page is not dimmed out and users can interact with the rest of the page. This also implies that the tab focus can move outside the dialog when it reaches the last focusable element.
+   *
+   * `alert`: is a special type of modal dialogs that interrupts the user's workflow to communicate an important message or ask for a decision. Unlike a typical modal dialog, the user must take an action through the options given to dismiss the dialog, and it cannot be dismissed through the dimmed background or escape key.
+   */
+  type?: 'modal' | 'non-modal' | 'alert';
+  /**
+   * Controls the open state of the dialog
+   * @default undefined
+   */
+  open?: boolean;
+  /**
+   * Default value for the uncontrolled open state of the dialog.
+   * @default false
+   */
+  defaultOpen?: boolean;
+  /**
+   * Callback fired when the component changes value from open state.
+   * @default undefined
+   */
+  onOpenChange?(event: MouseEvent | KeyboardEvent, data: DialogOpenChangeData): void;
+};
+
+type DialogOpenChangeData = {
+  /**
+   * The event source of the callback invocation
+   */
+  type: 'escapeKeyDown' | 'overlayClick' | 'triggerClick';
+  /**
+   * The next value for the internal state of the dialog
+   */
+  open: boolean;
+};
+```
+
+### DialogTrigger
+
+A non-visual component that wraps its child and configures them to be the trigger that will open or close a `Dialog`. This component should only accept one child.
+
+In case the trigger is used outside `Dialog` component it'll still provide basic [ARIA related attributes](#aria-roles-and-states) to it's wrapped child, but it won't be able to alter the dialog `open` state anymore, in that case the user must provide a [`controlled state`](#controlled-dialog)
+
+```typescript
+export type DialogTriggerProps = {
+  /**
+   * Explicitly declare if the trigger is responsible for opening,
+   * closing or toggling a Dialog visibility state.
+   * @default 'toggle'
+   */
+  type?: 'open' | 'close' | 'toggle';
```

Comment:
Do you have a scenario for `open` vs `toogle`? When `open` is needed?

I was thinking about `DialogTrigger` with toggle behavior as other components and `DialogCloseTrigger` (_ugly name_).

---

Again it might not be a good idea to use `type` as a prop name there.

Reply:
So, your suggestion would be a new compound component for the specific case of closing and keeping `DialogTrigger` as toggle.

I don't see a huge impact difference in supporting both scenarios in a single component instead though... I think for a moment I'll stick with a single component for this purpose, as they in the end do the same thing: they change open state.

As for the name of the props, how about `triggerType` or perhaps 3 boolean properties: `open`, `close`, `toggle`?

Reply:
`action` sounds better than `type` indeed.